### DISCUSSION TITLE
[20 Marzo 2023] Pipeline

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#80] [Pipeline (M-Z)]
Il programma qui di seguito effettua la somma dei valori (in $s0) degli elementi presenti nell’array A di lunghezza L, dopo averne modificato i valori negativi moltiplicandoli per 2. Se il risultato è maggiore o uguale a 10, viene stampata la stringa S.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione. 
Si indichino (ignorando hazard che possano concernere la syscall):
$1)$ le istruzioni tra le quali sono presenti **data hazard** – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
$2)$ le istruzioni tra le quali sono presenti **control hazard** – indicare i numeri di istruzione N ed M (nel formato N / M);
$3)$ quanti cicli di clock sono necessari ad eseguire il programma tramite **forwarding**, spiegando il calcolo effettuato;
$4)$ quanti cicli di clock sarebbero necessari ad eseguire il programma **senza forwarding**, spiegando il calcolo effettuato;
$5)$ quali sono, per ognuna delle cinque fasi, le **istruzioni** (o le bolle) in pipeline durante **il 13° ciclo di clock** (con **forwarding**);
$6)$ in quale fase la decisione delle istruzioni di jump sia compiuta nel diagramma qui sotto, e quante fasi di stallo causi potenzialmente ogni control hazard.
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/1ed2dbac-8195-406a-94e2-3af8751e9baf)
![Architettura Pipeline](https://github.com/Elia-Belli/ae2223/assets/107944476/4f230aed-9dc2-4ed9-b2ff-c9a0c0fb4960)
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11666095/AE.2021-22.-.Appello.S2.-.2023-03-20.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf)

**Comment by CasuFrost** [comment]
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/97389762/9346d92f-1719-4dfc-845b-e8b5c43c425a)
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/97389762/2f7df538-67e7-4b9b-9890-900b5c4a9204)
### Domanda 6 : 
La decisione di jump viene decisa in fase di Istruction Fetch, da come si può notare, durante questa fase,
i bit dell'istruzione [31-26] finiscono in un comparatore (con 2) per poi finire come selettero del mux, 
se tali bit sono uguali a 2 (opCode dell'istruzione jump) il mux selezionerà appunto l'indirizzo indicato 
dai bit [25-0] , è quindi nella fase di Fetch che viene deciso che verrà eseguito un Jump.

**Comment by Frisbon** [comment]
As sempre,
In formato PDF in modo che sia il piú leggibile possibile.

[**_La parte piú difficile is always la 6_**]
(https://github.com/sapienzastudentsnetwork/archelab2223/files/11873656/La.parte.piu.difficile.is.always.la.6.pdf)

https://youtu.be/So5UCib8jWU

**Reply by matypist** [reply]
Note per chi leggerà questa soluzione in futuro:
- Tra le istruzioni 29 e 31 non c'è un data hazard visto che entrambe le istruzioni insistono sì su **$v0** ma come solo output:
  29: addi **$v0**, $zero, 4
  30: syscall
  31: addi **$v0**, $0, 10
- Nel punto **n. 5** la risposta è corretta ma la domanda d'esame chiedeva del **13°** ciclo di clock
- Nel punto **n. 6**, per via della considerazione correttamente espressa in merito al Jump, un Control Hazard causerà 0 stalli nel caso del jump o al massimo 1 nel caso di Control Hazard in un'istruzione di branch anticipato in fase di ID

**P.S.:** molto utile il video, grazie per la condivisione!
